edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/MetadataStore.cs

```diff
@@ -0,0 +1,115 @@
+            this.edgeProductInfo = Preconditions.CheckNotNull(edgeProductInfo, nameof(edgeProductInfo));
+        }
+
+        public async Task<Option<ConnectionMetadata>> GetMetadata(string id)
+        {
+            Option<string> value = await this.metadataEntityStore.Get(id);
+            return await value.Match(
+                async v =>
+                {
+                    try
+                    {
+                        return Option.Some(JsonConvert.DeserializeObject<ConnectionMetadata>(v));
+                    }
+                    catch (JsonSerializationException)
+                    {
+                        // If deserialization fails, assume the string is an old productInfo.
+                        // We must do this only for migration purposes, since this store used to just be a productInfoStore.
+                        ConnectionMetadata metadata = new ConnectionMetadata() { ProductInfo = v };
+                        await this.SetMetadata(id, metadata);
+                        return Option.Maybe(metadata);
+                    }
+                },
+                () =>
+                {
+                    return Task.FromResult(Option.None<ConnectionMetadata>());
+                });
+        }
+
+        public async Task<Option<string>> GetModelId(string id)
+        {
+            Option<string> value = await this.metadataEntityStore.Get(id);
+            return value.Match(
+                v =>
+                {
+                    try
+                    {
+                        return Option.Some(JsonConvert.DeserializeObject<ConnectionMetadata>(v).ModelId);
+                    }
+                    catch (JsonSerializationException)
+                    {
+                        // If deserialization fails, assume the string is an old productInfo.
+                        return Option.None<string>();
+                    }
+                },
+                () => Option.None<string>());
+        }
+
+        public async Task<string> GetProductInfo(string id)
+        {
+            Option<string> value = await this.metadataEntityStore.Get(id);
+            return await value.Match(
+                async v =>
+                {
+                    try
+                    {
+                        return JsonConvert.DeserializeObject<ConnectionMetadata>(v).ProductInfo;
+                    }
+                    catch (JsonSerializationException)
+                    {
+                        // If deserialization fails, assume the string is an old productInfo.
+                        // We must do this only for migration purposes, since this store used to just be a productInfoStore.
+                        ConnectionMetadata metadata = new ConnectionMetadata() { ProductInfo = v };
+                        await this.metadataEntityStore.Put(id, JsonConvert.SerializeObject(metadata));
+                        return v;
+                    }
```

Comment:
Seems like this can be refactored into a helper along the line of GetOrMigrateProductInfo()